crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs

```diff
@@ -117,6 +118,80 @@ impl ArrayHeap {
         self.staged.insert(index, (value, op));
     }
 
+    /// This helper function transforms an expression into a linear expression, by generating a witness if the input expression is not linear nor constant
+    fn normalize_expression(expr: &Expression, evaluator: &mut Evaluator) -> Expression {
+        if !expr.is_linear() || expr.linear_combinations.len() > 1 {
+            let w = evaluator.create_intermediate_variable(expr.clone());
+            Expression::from(w)
+        } else {
+            expr.clone()
+        }
+    }
+
+    /// Decide which opcode to use, depending on the backend support
+    fn acir_gen(
+        &self,
+        evaluator: &mut Evaluator,
+        array_id: ArrayId,
+        array_len: u32,
+        is_opcode_supported: IsOpcodeSupported,
+    ) {
+        let dummy = MemoryBlock { id: AcirBlockId(0), len: 0, trace: Vec::new() };
+
+        if is_opcode_supported(&AcirOpcode::ROM(dummy.clone())) {
+            // If the backend support ROM and the array is read-only, we generate the ROM opcode
+            if matches!(self.typ, ArrayType::ReadOnly(_)) {
+                self.add_rom_opcode(evaluator, array_id, array_len);
+                return;
+            }
+        }
+        if !is_opcode_supported(&AcirOpcode::Block(dummy.clone())) {
+            if is_opcode_supported(&AcirOpcode::RAM(dummy)) {
+                self.add_ram_opcode(evaluator, array_id, array_len);
+            } else {
+                self.generate_permutation_constraints(evaluator, array_id, array_len);
+            }
+        } else {
+            self.add_block_opcode(evaluator, array_id, array_len);
+        }
```

Comment:
```suggestion
        if is_opcode_supported(&AcirOpcode::ROM(dummy.clone())) {
            // If the backend support ROM and the array is read-only, we generate the ROM opcode
            if matches!(self.typ, ArrayType::ReadOnly(_)) {
                self.add_rom_opcode(evaluator, array_id, array_len);
                return;
            }
        }

        if is_opcode_supported(&AcirOpcode::RAM(dummy.clone())) {
            self.add_ram_opcode(evaluator, array_id, array_len);
        } else if is_opcode_supported(&AcirOpcode::Block(dummy)) {
           self.add_block_opcode(evaluator, array_id, array_len);
        } else {
            self.generate_permutation_constraints(evaluator, array_id, array_len);
        }
```

We can flatten these last if-elses afaict.

Reply:
I have re-written it in the order of the opcode preference